Fix hit error meters not updating visual state when hidden

peppy opened the pull request:
It is an expectation of users that when the HUD is shown after a period of being hidden, it will visually reflect the state based on recent judgements.

To achieve this, I've added `AlwaysPresent` and moved the transform application to the meter level, rather than at a child level. If this is seen as a bad direction, `AlwaysPresent` can be applied to the drawable children and the transforms can be moved back.

Also of note, `ColourHitErrorMeter` is pretty weird. The flow class could potentially be removed and reduce `AlwaysPresent` usage by one. Can do that refactor as part of this PR if preferred.

Closes #18624.

peppy committed:
Add failing test coverage showing hit meters don't update when not visible

peppy committed:
Fix hit error meters not updating visual state when hidden

It is an expectation of users that when the HUD is shown after a period
of being hidden, it will visually reflect the state based on recent
judgements.

To achieve this, I've added `AlwaysPresent` and moved the transform
application to the meter level, rather than at a child level. If this is
seen as a bad direction, `AlwaysPresent` can be applied to the drawable
children and the transforms can be moved back.

Also of note, `ColourHitErrorMeter` is pretty weird. The flow class
could potentially be removed and reduce `AlwaysPresent` usage by one.
Can do that refactor as part of this PR if preferred.

Closes #18624.

peppy commented on osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs:
```diff
@@ -420,12 +423,26 @@ protected override void OnNewJudgement(JudgementResult judgement)
+
+            judgementsContainer.Add(judgementLine);
+
+            // Importantly, transforms should be applied in this method rather than constructed drawables
+            // to ensure that they are applied even when the `HitErrorMeter` is hidden (see `AlwaysPresent` usage).
```
I may have been doing something wrong in the earlier tests, I'll revisit this. It looked initially like children weren't being `LoadComplete`d even though the `HitErrorMeter` was set to `AlwaysPresent`.

peppy replied:
Reverted all unnecessary changes. Some of them felt like nice clean-up but probably best to keep this focused and simple.

peppy committed:
Revert most unnecessary changes

Turns out `AlwaysPresent` at top level is actually enough.

This reverts commit 86163d2225157af9c874af5b2dd59208572f8349.

frenzibyte commented:
While this fixes `HitErrorMeter` not updating when hidden by itself, it still doesn't update when combined with the `HUDOverlay`, and which I confirmed via logging on `HitErrorMeter.Update()` and going to a player session:

https://user-images.githubusercontent.com/22781491/173461407-0e839b01-73fe-479c-bddf-a64aa3be05be.mp4

Reason would be that the layer which hides in HUD overlay is the `MainHUDComponents` container, and once it's hidden and `IsPresent = false`, its children (HUD components, including `HitErrorMeter`) [cannot possibly update](https://github.com/ppy/osu-framework/blob/7610d868ead7f86e2606ba6b3f23b3c5b2605648/osu.Framework/Graphics/Containers/CompositeDrawable.cs#L906-L908).

Easiest fix I can think of would be adding `AlwaysPresent` to the HUD components container, but would imagine the components hiding individually in the future to avoid this, maybe.

frenzibyte reviewed (changes requested):
As noted.

peppy commented:
I tested in gameplay and it worked for me correctly. Maybe a regression with the last commit, try without that?

frenzibyte commented:
Can still see the issue after reverting last commit:

https://user-images.githubusercontent.com/22781491/173472699-8d5f6baa-0324-45b7-a7c9-8db82f8597cf.mp4

I believe this sort of issue may fix itself probably by having `HitErrorMeter` [not rely on scheduling](https://github.com/ppy/osu/blob/0771265cafbd0db82403f7e95cff892906a1bbfc/osu.Game/Screens/Play/HUD/HitErrorMeters/HitErrorMeter.cs#L46-L47) (which should be the reason why all judgements queue up to appear at the point of displaying meter again). Haven't tested that out myself, but just putting it here for visibility.

peppy commented:
Are you actually *looking at the display*? It doesn't matter if update is running or not, the important thing is whether it behaves correctly.

peppy committed:
Ensure containers which are being used to hide HUD elements still update for now

I don't think this is necessarily a final solution (as this means all
HUD elements are adding overhead even when not visible), but this will
make the implementations much easier for the time being.

I've checked and can't notice any perceivable overhead in profiling so
we should be fine for now.

peppy commented:
I've gone with the `AlwaysPresent` solution for now as it's the simplest, and will potentially cover other HUD elements that require similar updates.

Ensure containers which are being used to hide HUD elements still update for now

I don't think this is necessarily a final solution (as this means all HUD elements are adding overhead even when not visible), but this will make the implementations much easier for the time being.

I've checked and can't notice any perceivable overhead in profiling so we should be fine for now.

peppy committed:
Add test coverage of HUD components still getting updated when hidden

peppy commented on osu.Game/Screens/Play/HUDOverlay.cs:
```diff
@@ -108,6 +111,7 @@ public HUDOverlay(DrawableRuleset drawableRuleset, IReadOnlyList<Mod> mods)
                     Margin = new MarginPadding(10),
                     Spacing = new Vector2(10),
                     AutoSizeAxes = Axes.Both,
+                    AlwaysPresent = true,
```
Yes, you are correct in thinking that.

peppy committed:
Fix `AlwaysPresent` specification in wrong container

frenzibyte committed:
Merge branch 'master' into fix-hit-error-when-not-visible

frenzibyte committed:
Merge branch 'master' into fix-hit-error-when-not-visible
